src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -898,6 +878,138 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
             return dictionary;
         }
 
+        internal static void AddToMultiValueDictionaryBuilder<K, T>(Dictionary<K, object> accumulator, K key, T item)
+            where K : notnull
+            where T : notnull
+        {
+            if (accumulator.TryGetValue(key, out var existingValueOrArray))
+            {
+                if (existingValueOrArray is ArrayBuilder<T> arrayBuilder)
+                {
+                    // Already a builder in the accumulator, just add to that.
+                }
+                else
+                {
+                    // Just a single value in the accumulator so far.  Convert to using a builder.
+                    arrayBuilder = ArrayBuilder<T>.GetInstance(capacity: 2);
+                    arrayBuilder.Add((T)existingValueOrArray);
+                    accumulator[key] = arrayBuilder;
+                }
+
+                arrayBuilder.Add(item);
+            }
+            else
+            {
+                // Nothing in the dictionary so far.  Add the item directly.
+                accumulator.Add(key, item);
+            }
+        }
+
+        internal static void CreateNameToMembersMap
+            <TNamespaceOrTypeSymbol, TNamedTypeSymbol, TNamespaceSymbol>
+            (Dictionary<string, object> dictionary, Dictionary<String, ImmutableArray<TNamespaceOrTypeSymbol>> result)
+            where TNamespaceOrTypeSymbol : class
+            where TNamedTypeSymbol : class, TNamespaceOrTypeSymbol
+            where TNamespaceSymbol : class, TNamespaceOrTypeSymbol
+        {
+            foreach (var kvp in dictionary)
+            {
+                object value = kvp.Value;
+                ImmutableArray<TNamespaceOrTypeSymbol> members;
+
+                var builder = value as ArrayBuilder<TNamespaceOrTypeSymbol>;
+                if (builder != null)
+                {
+                    Debug.Assert(builder.Count > 1);
+                    bool hasNamespaces = false;
+                    for (int i = 0; (i < builder.Count) && !hasNamespaces; i++)
+                    {
+                        if (builder[i] is TNamespaceSymbol)
+                        {
+                            hasNamespaces = true;
+                            break;
+                        }
+                    }
+
+                    members = hasNamespaces
+                        ? builder.ToImmutable()
+                        : ImmutableArray<TNamespaceOrTypeSymbol>.CastUp(builder.ToDowncastedImmutable<TNamedTypeSymbol>());
+
+                    builder.Free();
```

Comment:
calling `ToImmutable` followed by `Free` dosen't allow for the optimization of moving the internal array to the immutable array without a copy that ToImmutableAndFree allows for.

Reply:
> calling ToImmutable followed by Free dosen't allow for the optimization of moving the internal array to the immutable array without a copy that ToImmutableAndFree allows for.

This is an "extract common code" refactoring, I am not interested in changing what the code is doing and how it does that.

Reply:
i'm pointing out that the code you're extracting out has issues with it.  Since you are working with the code *now*, it's a good time to actually address the problems in it.

Reply:
> i'm pointing out that the code you're extracting out has issues with it.

This is definitely something to consider as a future improvement, I will open a follow-up issue.

> Since you are working with the code now, it's a good time to actually address the problems in it.

I do not agree with that. I prefer to separate the concerns.

Reply:
> I will open a follow-up issue.

#67990